components/channel/channel.c

```diff
@@ -0,0 +1,178 @@
+static bool receiver_recv(struct datadog_php_receiver_s *receiver, void **data, uint64_t timeout_nanos) {
+    if (!receiver || !receiver->channel) {
+        return false;
+    }
+
+    datadog_php_channel_impl *channel = receiver->channel;
+    uv_mutex_lock(&channel->mutex);
+    datadog_php_queue *queue = &channel->queue;
+    bool succeeded = queue->try_pop(queue, data);
+
+    // If there wasn't an item but there are known producers, wait and try again
+    if (!succeeded && channel->sender_count && timeout_nanos) {
+        uint64_t now = uv_hrtime();
+        uint64_t timeout_target = now + timeout_nanos;
+        do {
+            // Handle being signalled, waking up spuriously, and timing out the same.
+            (void)uv_cond_timedwait(&channel->condvar, &channel->mutex, timeout_target - now);
+            succeeded = queue->try_pop(queue, data);
+            if (succeeded) {
+                break;
+            }
+
+            now = uv_hrtime();
+        } while (channel->sender_count && now < timeout_target);
+    }
+
+    uv_mutex_unlock(&channel->mutex);
+    return succeeded;
+}
+
+static void receiver_dtor(struct datadog_php_receiver_s *receiver) {
+    if (receiver && receiver->channel) {
+        datadog_php_channel_impl *channel = receiver->channel;
+        receiver->channel = NULL;
+        uv_mutex_lock(&channel->mutex);
+        // todo: underflow checks
+        bool dtor = !(--channel->receiver_count | channel->sender_count);
+        uv_mutex_unlock(&channel->mutex);
+
+        // Cannot dtor the mutex while it is held.
+        if (dtor) {
+            channel_dtor(channel);
+        }
+    }
+}
+
+static bool sender_send(struct datadog_php_sender_s *sender, void *data) {
+    if (!sender || !sender->channel) {
+        return false;
+    }
+
+    datadog_php_channel_impl *channel = sender->channel;
+    uv_mutex_lock(&channel->mutex);
+    datadog_php_queue *queue = &channel->queue;
+    bool sent = queue->try_push(queue, data);
+    if (sent) {
+        uv_cond_signal(&channel->condvar);
+    }
+    uv_mutex_unlock(&channel->mutex);
+    return sent;
+}
+
+static void sender_dtor(struct datadog_php_sender_s *sender) {
+    if (sender && sender->channel) {
+        datadog_php_channel_impl *channel = sender->channel;
+        sender->channel = NULL;
+        uv_mutex_lock(&channel->mutex);
+        // todo: underflow check
+        bool dtor = !(channel->receiver_count | --channel->sender_count);
+        uv_mutex_unlock(&channel->mutex);
+
+        // Cannot dtor the mutex while it is held.
+        if (dtor) {
+            channel_dtor(channel);
+        }
+    }
+}
+
+static bool sender_clone(datadog_php_sender *self, datadog_php_sender *clone) {
+    if (!self || !clone || !self->channel) {
+        return false;
+    }
+
+    datadog_php_channel_impl *channel = self->channel;
+    uv_mutex_lock(&channel->mutex);
+    // todo: overflow checks
+    ++channel->sender_count;
+    uv_mutex_unlock(&channel->mutex);
```

Comment:
It doesn't matter because `receiver_recv` will unlock the mutex while it waits, and re-acquire it when it resumes; the period where the lock is contended should be short.